### ISSUE TITLE
basket doesn't save on logout

### ISSUE THREAD

**Issue by NickWeir63** [openfoodfoundation/openfoodnetwork#1031]
Stroudco shoppers want to be able to leave items in their basket that are not lost when they log out so they can add items during the week and then review and confirm them just before the OC closes

**Comment by lin-d-hop** [comment]
@stveep Would you be willing to look into this issue a little more? Perhaps to scope it out a little more for @lebreeze.

**Comment by lin-d-hop** [comment]
A little more specification from a non-technical perspective:

As a shopper, when I am logged in I would like my cart to persist for the hub I am shopping with. I would like OFN to remember the last cart that I created such then if I log out and log back in again the last cart is persisted.

If I change enterprises, order cycles or the order cycle closes my cart should not persist.

Currently if a user is shopping with one enterprise, then they switch to another enterprise their cart is cleared from the scope. This behaviour is enables multiple shopfronts to exist simultaneously. However the majority of shoppers only shop from a single hub. Therefore there is a clear benefit for a customers cart to persist if they log out and log back in to shop at the same hub.

The spree_order model enables carts to be persisted to a user and spree_line_items are also persisted. So to my understanding this issue is pertaining to reloading the cart to the scope when a user logs back in, with some checks to ensure the order cycle has not closed.

I think this issue will need a couple hours of scoping first to fully understand the implications. @stveep Do you think this might be suitable for @lebreeze to ldelve further into?

@oeoeaio @mkllnk If this triggers any alarm bells to you would you mind noting them here too?

**Comment by mkllnk** [comment]
I think that would be great. There was once a multicart feature before my time. I'm not sure if that actually worked. I definitely find the current cart system confusing and brittle. But it might be a bit difficult to work around Spree's model of only one cart.

**Comment by oeoeaio** [comment]
This would be different to a multi-cart, right? We would simply be tweaking the `current_order` method in `lib/spree/core/controller_helpers/order_decorator.rb` to make it search the database for a relevant incomplete order, in the event that there is no `order_id` in the session.

If this search was restricted to incomplete orders owned by the current user, for the current shop and in the current order cycle, I can't see a problem with it.

**Comment by kirstenalarsen** [comment]
the multi-cart never worked. it was a folly, that was abandoned ass the complexity became too much :)

**Comment by lin-d-hop** [comment]
While multi-cart would be amazing, this enhancement falls a long way short. Thanks for the clear spec @oeoeaio.

Tagging @levent as I tagged him incorrectly above.

**Comment by NickWeir63** [comment]
I dont' think @levent is getting these notifications - when I spoke to him just now he has not seen these github notifications.  is there something wrong with his membership of this group?

**Comment by levent** [comment]
Hi Nick, notifications are fine. It was just the incorrect username being used previously.
Cheers

**Comment by levent** [comment]
Based on @lin-d-hop and @oeoeaio's comments this doesn't seem too complicated. I will familiarise myself with the cart logic today and get back to you.

**Comment by levent** [comment]
Ok. I went on a bit of a roundabout tour and discovered that there were some changes made a while ago that modified the cart logic.

Out of the box, spree attempts to merge the current session's cart with any incomplete cart when a user logs in.

In October 2012 two commits https://github.com/openfoodfoundation/openfoodnetwork/commit/27b7d59f460fa3440a6c7fd90f3408929f0f415d#diff-af734c2388f7722b392fa0d4834437a4R12 and then https://github.com/openfoodfoundation/openfoodnetwork/commit/568d948cbe123bc98228566a240cbd7e43eb4a4d were added to firstly check that the carts were only merged if from the same distributor and secondly to delete any incomplete carts that didn't match the distributor.

In August 2013 this was again simplified to just delete any incomplete carts and not bother merging at all (commit: https://github.com/openfoodfoundation/openfoodnetwork/commit/5fac29d2ed91577fd66bccc5fc49caa25aed93ed) This was done to fix https://github.com/openfoodfoundation/openfoodnetwork/issues/20.

Long story short, removing the spree override of ```set_current_order``` https://github.com/openfoodfoundation/openfoodnetwork/compare/master...levent:1031-spike-remove-override?expand=1 will mean that all tests pass and logged out user's cart items are merged with any previous cart items they had last time they were logged in. It won't handle jumping from shop to shop yet.

@RohanM do you remember the rationale behind the overrides to ```set_current_order``` and what regression we would introduce by removing / reverting it? Unfortunately any specs written around that logic seem to have been lost over the years during major refactoring or redesign stages.

I also found a small sql query bug along the way where the ```last_incomplete_order``` method would always return an empty result if the user didn't have any ```account_invoices```. If it's common *not* to have account_invoices then the app will always think the user does not have any incomplete orders. I can create a pull request from this fix: https://github.com/openfoodfoundation/openfoodnetwork/compare/master...levent:1031-bugfix?expand=1

Notes for my benefit on this: https://gist.github.com/levent/a7c5e1e78db0c0e6d5a337459629a848

**Comment by lin-d-hop** [comment]
Thanks @levent!
@daniellemoorhead would you mind following up @levents question above with @rohanm?

**Comment by RohanM** [comment]
Hi @levent,

Nice work researching an obscure issue! I think the rationale is sitting in the commit message of https://github.com/openfoodfoundation/openfoodnetwork/commit/568d948cbe123bc98228566a240cbd7e43eb4a4d. It was very common for users to have incomplete orders sitting around in the database. When they completed checkout, the system was restoring their cart from their incomplete order and they'd end up with a bunch of items mysteriously appearing in their cart. I'm thinking that with a bit of testing we could ensure we don't introduce a regression around this.

Hope that helps, let me know if you have any other questions.

**Comment by lin-d-hop** [comment]
Hi all, Levent pointed out that this issue needs more clarification... 
@levent @oeoeaio Can you add any cases that we need a case for to the list below? Thanks!
@OliverUK @sstead Would you comment on the below also? In particular the behaviour in case 2.1. Thanks :-)

As customer I create a basket then log out:
  1. I log back in without creating a new basket:
  1.1 Order cycle of last basket is still open -> Old Basket Restored
  1.2 Order cycle of last basket is now closed -> Old Basket Deleted
  2 I create a new basket in the same shop then log back in:
  2.1 Order cycle of last basket is open -> Old Basket Added to New basket
  2.2 Order cycle of last basket is closed -> Old Basket Replaced by New basket
  3. I create a new basket in a different shop then log back in:
  3.1 Order cycle of last basket is open -> Basket Replaced by New basket
  3.2 Order cycle of last basket is closed -> Basket Replaced by New basket

In the case of the old basket being restored, it is likely that some of the products previously added to the list may now be sold out. In this case I would suggest these products quietly disappear. @levent can you comment on the legacy behaviour in this case?

**Comment by NickWeir63** [comment]
i have talked to one of the shoppers who first raised this.  she think that:

- in case 2.1 (above) the shopper is likely to have forgotten that they have items in their basket and there should be either be a warning that they are about to combine the baskets or an option either to combine baskets or delete the previous basket

- In the case of the old basket being restored, if some of the products previously added to the basket are now sold out, then there should be a warning to say that these products have been deleted from the basket.

Sorry - I expect these are tricky demands!

**Comment by OliverUK** [comment]
Hi @lin-d-hop 
I wonder if it's worth the hassle letting people do this, considering the implications on limited stock items. I too think that if in 2.1 the old and new baskets get merged, this would cause confusion and would need a warning or the option to add or discard the old items. 

I personally would guess most people shop in one session.

**Comment by oeoeaio** [comment]
I agree with @OliverUK.

My feeling is that 2.1 will be annoying more often than it is helpful, with the possible exception of the 'new' basket being empty. If a customer is already shopping and then logs in, my preference would be to just keep their basket as is, without looking for previous baskets. I think 1.1 and 3.1 are ok, but as mentioned we will need to notify customers when stock levels mean their existing order can no longer be fulfilled. 

There is a method called something like cap_quantity_at_stock! on LineItem that we can use for this, then we just need to track any changes that were made and notify the customer. I am doing a similar thing with standing orders.

1.2, 2.2 and 3.2 are existing behaviour as far as I understand. I think they're ok.

**Comment by levent** [comment]
@lin-d-hop if you make the following changes to restore the default merge basket functionality:

A) This bug fixed: [BUG: Incomplete cart query always empty](https://github.com/openfoodfoundation/openfoodnetwork/compare/master...levent:1031-bugfix?expand=1)
B) Spree default behaviour restored: [Remove override that destroys incomplete orders](https://github.com/openfoodfoundation/openfoodnetwork/compare/master...levent:1031-spike-remove-override?expand=1)

* **NOTE: This code knows nothing about order cycles or enterprises**

> In the case of the old basket being restored, it is likely that some of the products previously added > to the list may now be sold out. In this case I would suggest these products quietly disappear.
> @levent can you comment on the legacy behaviour in this case?

The baskets are initially merged (including out of stock inventory) and then on a subsequent request the user is informed via a modal dialogue:

![screenshot 2016-11-24 09 41 21](https://cloud.githubusercontent.com/assets/2386/20593682/342b0328-b22b-11e6-94de-be94db7c5ad0.png)

**Comment by levent** [comment]
I am currently working on a branch that satisfies:

```
A) If user already has items in their basket on log in,
     do not modify the basket.
B) If user does **not** have items in their basket on login,
     use the old basket if it's the same distributor and order cycle
```


This would/should mean:

```
1.1 True - old basket restored
1.2 True - continue with empty basket
2.1 False - continue with basket used whilst logged out
```

I don't fully understand the user journey for 2.2, 3.1 and 3.2 but if you understand how spree determines the last_incomplete_order and follow A) and B) above, you may get the answer.

### What is last_incomplete_order?

The most recently created, incomplete cart (across the entire system) associated with the user.

### Am I oversimplifying or is this a useful and reasonable improvement?

**Comment by lin-d-hop** [comment]
Thanks everyone. 
Based on comments above I agree with y'all that 2.1 should be updated to use new basket. 

So, in that case, @Levent, your conditions sound good and I agree that your changes are a useful and reasonable improvement.
```
A) If user already has items in their basket on log in,
     do not modify the basket.
B) If user does **not** have items in their basket on login,
     use the old basket if it's the same distributor and order cycle
```

@levent, your notification looks good to me. @oeoeaio is this consistent with the communication of out of stock items in SO?

**Comment by lin-d-hop** [comment]
Hi @levent 
With the two fixes you made above is this ready for review and testing? Or is there something you are still working on here?

**Comment by levent** [comment]
Hi @lin-d-hop 
Not quite, I'm afraid.
Will be ready tomorrow.

**Comment by lin-d-hop** [comment]
Great thanks @levent! Just wanted to check this wasn't waiting on me :-)

**Comment by OliverUK** [comment]
Happy to test but can't find in buildkite, @lin-d-hop .

**Comment by lin-d-hop** [comment]
@OliverUK Now has a build so over to you :-)

**Comment by OliverUK** [comment]
@lin-d-hop 
![screen shot 2017-01-22 at 10 07 11](https://cloud.githubusercontent.com/assets/17590921/22181772/b255df24-e08a-11e6-9ab0-589665cf29de.png)

**Comment by lin-d-hop** [comment]
Try again @OliverUK. It's on staging now.

**Comment by OliverUK** [comment]
@lin-d-hop Hi Lynne, is this the right build?
![screen shot 2017-01-29 at 12 15 28](https://cloud.githubusercontent.com/assets/17590921/22403883/c14b6124-e61c-11e6-8f47-79ecf4eaa63e.png)


It was already staged but didn't seem to have an effect so I rebuilt and redeployed to staging but when I try putting things in my shopping basket, log out and log back in, the basket is empty. If I log in before putting anything in, OFN should remember the old basket but it doesn't.

**Comment by lin-d-hop** [comment]
Deploying this doesn't have any determinable affect... so this needs some looking into. Likely a quick one if anyone wants to take a look?

**Comment by lin-d-hop** [comment]
@sstead Would be good to see if this feature is ready for merge. It has been tested by @OliverUK. 

He spotted a bug unrelated to this issue, it sounds familiarish but I'm not sure if it has already been clocked.

**Comment by sstead** [comment]
Testing notes:
All use cases working as desired...
1.	I log back in without creating a new basket:
1.1 Order cycle of last basket is still open -> Old Basket Restored YES
1.2 Order cycle of last basket is now closed -> Old Basket Deleted YES- and I get told the OC had closed.
2.	I create a new basket in the same shop then log back in:
2.1 Order cycle of last basket is open -> Old Basket Added to New basket NO- your new basket overrides the old one.
2.2 Order cycle of last basket is closed -> Old Basket Replaced by New basket YES
3.	I create a new basket in a different shop then log back in:
3.1 Order cycle of last basket is open -> Basket Replaced by New basket YES
3.2 Order cycle of last basket is closed -> Basket Replaced by New basket YES

One glitch, this may be the same one picked up by Oliver...
If I had an item in my cart, then I logged out, then the item was removed from the OC (either just that item or the supplier), then I log in and go to the shopfront- my cart hangs and I'm unable to proceed through checkout. If I instead login and click straight on the edit my cart dropdown link, I can proceed through checkout and purchase something that's no longer in the OC.

**Comment by lin-d-hop** [comment]
@OliverUK can you confirm if this is the same issue you spotted? If so this issue was not introduced by this bug but exists already in production.

**Comment by mkllnk** [comment]
Sally tested production and said that this issue isn't there, because carts don't persist.

**Comment by OliverUK** [comment]
@lin-d-hop I does sound like the same issue although I didn't get where the cart hangs. But I didn't try removing things from the OC but rather make the item "sell out". For me this led to being able to complete checkout and the items had negative stock.

@mkllnk I don't know if I'm understanding you correctly but I do think the issue exists on production. If a customer does NOT log out and back in but completes his order in one session, it's still possible that an items will sell out before he or she completes checkout. I tried this on UK production and for me the result was the same as above: I can order and the stock level goes negative.

**Comment by mkllnk** [comment]
@sstead Can you reproduce the error like @OliverUK describes it?

I haven't tested this. How difficult is it to fix this issue?

**Comment by lin-d-hop** [comment]
I also haven't tested this, but sounds like the most likely scenario is that both sally and oliver are right.

This enhancement thus significantly increases the likelihood that the issue will be experienced by a user and therefore it becomes a priority to resolve at least for this case.

**Comment by daniellemoorhead** [comment]
Hi all, just checking where we are with this. It's labelled as needing Aus Review, is this still the case? @lin-d-hop @NickWeir63

**Comment by OliverUK** [comment]
@daniellemoorhead I believe the issue of saving the basket is all ok but I've created a new issue #1491 for the separate problem of being able to check out items beyond available stock.

**Comment by lin-d-hop** [comment]
Thanks @OliverUK

**Comment by lin-d-hop** [comment]
@OliverUK 
Can you test this again but following the tip in #1335 first please?
Soo...
1) check inventory settings to confirm that Backordering in not enabled.
2) Repeat your testing to confirm if the order can still be made.

Thanks!

**Comment by OliverUK** [comment]
@lin-d-hop 
I don't usually get involved with super admin settings and didn't know this existed but I've done what you said and disabled backorders. This resulted in the second order for the same items getting bounced back to the cart with a message saying that one item in the order has become unavailable. 

As far as I could tell, everything else still behaves as experienced above. 

I take it these means backorders are also enabled on UK production? As far as I could find, it doesn't actually do anything useful. It doesn't put the item back in my basket once stock is replenished and when I log back in. Nor can I see anything in admin showing me backorders.

**Comment by lin-d-hop** [comment]
@OliverUK I didn't know it existed either so nay bother about that!
I've turned it off on UK production too as this is not a feature anyone has been using or wanting. Thanks for playing with it.

So as far as we can tell this issue is now complete and ready for AU testing and merge, yes?

**Comment by OliverUK** [comment]
LOL So we were all on a wild goose hunt, eh?
Agreed, should be good to go.

**Comment by lin-d-hop** [comment]
:-) Happens

**Comment by lin-d-hop** [comment]
Could I just confirm @OliverUK, did you test THIS enhancement? Did you test that if the item goes out of stock before the basket is _retrieved_ you can't make the purchase?

**Comment by OliverUK** [comment]
What exactly is meant to happen @lin-d-hop ?
I just had a quick go and it seems that the item simply doesn't appear in the basket when I log back in. Is that what we are looking for.

**Comment by lin-d-hop** [comment]
I guess it would be better if the user was informed, but at least they can't make the purchase.
Hrm...
@sstead We've discovered that as long as 'Allow Backordering' in Inventory settings is not set then the product just drops out of the basket when the user logs back in if it goes out of stock between the basket creation and the user ordering. Obviously this isn't the ideal functionality but would you be happy for this to pass testing and be merged?

**Comment by sstead** [comment]
I'm happy to retest this, but I predict this issue will still be a problem (I don't think it's related to the one above).

I've checked on Production (where backorders are disallowed), and the issue spotted in testing (below) is occurring on production. This means that if a customer is shopping, they have an item in their cart, and then the hub removes the item from the OC, they can still purchase it. The likelihood of this occurring with increase with the introduction of this feature.

_"If I had an item in my cart, then I logged out, then the item was removed from the OC (either just that item or the supplier), then I log in and go to the shopfront- my cart hangs and I'm unable to proceed through checkout. If I instead login and click straight on the edit my cart dropdown link, I can proceed through checkout and purchase something that's no longer in the OC."_

**Comment by lin-d-hop** [comment]
@jeronimo Could you take a look at this again please? See Sallys comment above, sounds like 'Allow backorder' is only one part of the issue. Robs comments above should give you direction on where to start to track it down. Thanks!

**Comment by OliverUK** [comment]
@sstead @lin-d-hop 
Did Lynne not turn off backorders after testing? So the problem was occurring while they were allowed?

**Comment by lin-d-hop** [comment]
Yep

⁣Sent from BlueMail ​

**Comment by jeronimo** [comment]
I have added two additional feature tests with the code @levent has submitted. In both cases (login back with incomplete order) it doesn't let to proceed from shopfront nor cart if there is insufficient stock when backorders are disallowed.
I tested it manually in my local dev environment as well as the code passed in Travis (though it allows backorders by default and I had to disable inside tests).

**Comment by lin-d-hop** [comment]
Would be great to dust this oldy off and see if we can move it along.....

**Comment by myriamboure** [comment]
Closing, wishlist item created here: https://community.openfoodnetwork.org/t/cart-should-save-on-log-out/1175
#gitcull2017

**Comment by lin-d-hop** [comment]
@myriamboure @oeoeaio @enricostano @sauloperez 
Given this issue is finished to the simple spec but is waiting on a bug that is an actual OFN bug I really don't want to close this and move it onto the wishlist.
Can we discuss this one?

**Comment by lin-d-hop** [comment]
@myriamboure It's not this issue that needs further inception. It's the solution to the bug #1491.

**Comment by myriamboure** [comment]
Sure @lin-d-hop, I would say if a bug is blocking a feature that has been developped and just can't be merged because of that bug, it makes sense to me to prioritize. It is part of "finishing what has been already done" ... but then let's agree on our first "prioritization ceremony" to put 1491 in delivery train backlog OR create an epic if needs inception... does it make sense? I switched from wishlist to icebox this feature item, so that we don't forget it but to avoid to have lots of issues open that we can't treat, I suggest to reopen the issue and PR when the bug is treated. Or else we can create an epic and in the inception, the bug resolution will be one of the stories? So we treat them all in the feature dev process. What do you think?

**Comment by lin-d-hop** [comment]
The best solution to this bug might be quite big so your proposal is a good one! I feel reassured that this won't become lost in the wishlist ocean. Thank you :-)
